### PR TITLE
BE-2102: Init for `cimp` library and cli.

### PR DESCRIPTION
**Pull request by kainobor**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,41 @@
+.PHONY: gen
+COVERAGE_FILE ?= coverage.txt
+
+#######################################################
```
Comment (mkorolyov) [inline Makefile]: there is no gitlab here.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+precompile:
+	go build ./...
+
+compile:
+ifeq (${BINARY_VERSION},unspecified)
+	$(eval BINARY_VERSION:=$(shell ./tag.sh -i))
+endif
+	go build \
+	-ldflags="-s -w \
+	-X 'pkg.humans.net/lib/grpc-core/version.Version=${BINARY_VERSION}'" \
```
Comment (mkorolyov) [inline Makefile]: ?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	-o bin/binary \
+	cmd/${BINARY_NAME}/main.go
+
+new_tag:
+	./tag.sh -p
```
Comment (mkorolyov) [inline Makefile]: also not needed

File: lib/cimp/fileformat.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	"fmt"
+	"path/filepath"
+
+	"github.com/go-yaml/yaml"
+)
+
+type FileFormat string
+
+const (
+	JsonFormat FileFormat = "json"
```
Comment (unkeep) [inline lib/cimp/fileformat.go]: JSON(YAML)Format

File: cmd/cimp/main.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import (
+	"flag"
+	"fmt"
+	"path/filepath"
+
+	"github.com/humans-group/consul-importer/lib/cimp"
+)
+
+func main() {
+	pathRaw := flag.String("p", "./config-local.yaml", "Path to yaml with config-file which should be imported")
```
Comment (mkorolyov) [inline cmd/cimp/main.go]: `./config-local.yaml` -> `config.yaml` ?

File: cmd/cimp/main.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	formatRaw := flag.String("f", "", "File format: json, yaml. If empty - got from extension. Default: yaml")
+	consulEndpoint := flag.String("c", "consul:8500", "Consul endpoint in format: `address:port`")
```
Comment (mkorolyov) [inline cmd/cimp/main.go]: consul:8500 -> 127.0.0.1:8500

File: cmd/cimp/main.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	path, err := filepath.Abs(*pathRaw)
+	check(err)
+
+	format, err := cimp.InitFormat(*formatRaw, path)
+	check(err)
+
+	kv := cimp.NewKV()
+	check(kv.FillFromFile(path, format))
+
+	serverName, err := kv.GetString("server/name")
```
Comment (mkorolyov) [inline cmd/cimp/main.go]: must be in the config, why general lib bounded to our internal setup?

File: lib/cimp/kv.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	case string:
+		result += curKey
+	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
+		result += fmt.Sprintf("%d", curKey)
+	case float32, float64:
+		result += fmt.Sprintf("%f", curKey)
+	case bool:
+		result += fmt.Sprintf("%t", curKey)
+	default:
+		panic(fmt.Sprintf("invalid config key type: %#v", curKey))
```
Comment (mkorolyov) [inline lib/cimp/kv.go]: add type info

**Review by mkorolyov — CHANGES_REQUESTED** [review]

File: lib/cimp/kv.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+	return "", fmt.Errorf("value %q: %w", key, ErrorTypeIncorrect)
+}
+
+func (kv KV) AddPair(pair api.KVPair) {
+	kv[pair.Key] = pair.Value
+}
+
+func (kv KV) AddPrefix(prefix string) {
+	newKV := NewKV()
```
Comment (unkeep) [inline lib/cimp/kv.go:62]: it's safe to delete a visited element from a loop, so we don't have to create one more map instance

File: lib/cimp/kv.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	return nil
+}
+
+func keyToString(prefix string, key interface{}) (string, error) {
+	var result string
+	if len(prefix) > 0 {
+		result = prefix + "/"
+	}
+
+	switch curKey := key.(type) {
```
Comment (unkeep) [inline lib/cimp/kv.go:104]: why not just `fmt.Sprint(key)` ?

File: lib/cimp/repo.go [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+type Config struct {
+	Address string
+}
+
+type ConsulRepo struct {
+	client *api.Client
+}
+
+func InitRepo(cfg Config) (*ConsulRepo, error) {
```
Comment (unkeep) [inline lib/cimp/repo.go]: mb Repo -> Storage? it's would be more natural for e.g consulStorage

File: lib/cimp/repo.go [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	}, nil
+}
+
+func (repo *ConsulRepo) Save(kv KV) error {
+	for k, v := range kv {
+		kvPair := &api.KVPair{
+			Key:   k,
+			Value: []byte(fmt.Sprintf("%v", v)),
+		}
+		_, err := repo.client.KV().Put(kvPair, nil)
```
Comment (unkeep) [inline lib/cimp/repo.go]: there is a batch method which would be more effective rather than per key put

**Review by unkeep — CHANGES_REQUESTED** [review]

File: lib/cimp/kv.go (hunk 9 above)
Comment (unkeep) [reply lib/cimp/kv.go:62]: never mind :)

Comment (unkeep) [reply lib/cimp/kv.go:62]: we cant insert in a loop

Comment (kainobor) [reply lib/cimp/kv.go:62]: yup, we can get keys "a/b" and "b". With prefix "a/" it'll be a problem

File: lib/cimp/kv.go (hunk 10 above)
Comment (kainobor) [reply lib/cimp/kv.go:104]: For more obvious code. Also, it can be configurable in future.

File: lib/cimp/repo.go (hunk 11 above)
Comment (kainobor) [reply lib/cimp/repo.go]: valid

File: lib/cimp/repo.go (hunk 12 above)
Comment (kainobor) [reply lib/cimp/repo.go]: > there is a batch method which would be more effective rather than per key put

fixed

File: lib/cimp/repo.go (hunk 11 above)
Comment (kainobor) [reply lib/cimp/repo.go]: > mb Repo -> Storage? it's would be more natural for e.g consulStorage

fixed

File: lib/cimp/fileformat.go (hunk 4 above)
Comment (kainobor) [reply lib/cimp/fileformat.go]: > JSON(YAML)Format

fixed

File: Makefile (hunk 1 above)
Comment (kainobor) [reply Makefile]: fixed

File: Makefile (hunk 2 above)
Comment (kainobor) [reply Makefile]: fixed too

File: Makefile (hunk 3 above)
Comment (kainobor) [reply Makefile]: fixed

File: cmd/cimp/main.go (hunk 5 above)
Comment (kainobor) [reply cmd/cimp/main.go]: fixed

File: cmd/cimp/main.go (hunk 6 above)
Comment (kainobor) [reply cmd/cimp/main.go]: fixed

File: cmd/cimp/main.go (hunk 7 above)
Comment (kainobor) [reply cmd/cimp/main.go]: > must be in the config, why general lib bounded to our internal setup?

added as cli-param

**Pull request merged**